packages/osd-agents/src/agents/langgraph/prompt_manager.ts

```diff
@@ -0,0 +1,335 @@
+
+You have access to the following tools through MCP (Model Context Protocol) servers:
+
+${toolDescriptions}
+
+${openSearchContext}
+
+Key behaviors:
+- Be concise and direct
+- Focus on the specific task at hand
+- Use available tools when appropriate
+- Provide clear, actionable responses
+- When asked about your tools, list ONLY the MCP tools shown above
+
+${this.getToolValidationRules()}`;
+  }
+
+  private getToolValidationRules(): string {
+    return `
+CRITICAL: Tool Parameter Validation and Self-Correction
+
+When using tools, you MUST follow this validation process:
+
+1. ANALYZE TOOL SCHEMA: Before calling any tool, carefully examine its description and parameter requirements
+   - Pay special attention to REQUIRED PARAMETERS listed in the tool description
+   - Review parameter types and constraints
+   - Ensure you understand what each parameter expects
+
+2. PARAMETER VALIDATION: Before making a tool call, verify:
+   - All required parameters are provided
+   - Parameter values are in the correct format
+   - No required parameters are null, undefined, or empty
+
+3. SELF-CORRECTION FLOW: If a tool call fails due to parameter validation:
+   - Immediately analyze the error message to identify missing or incorrect parameters
+   - Review the tool's parameter requirements again
+   - Ask the user for clarification if needed parameter values are not available in the context
+   - Retry the tool call with the correct parameters
+   - Do NOT make the same parameter mistake twice
+
+4. MULTI-TURN CORRECTION: For complex tool interactions:
+   - If you receive an error about missing parameters
+   - Stop and identify what information you need
+   - Either extract the required information from context or ask the user
+   - Retry with complete parameter set
+
+Example self-correction pattern:
+- Tool call fails: "Missing required parameter: path"
+- Response: "I need to provide the path parameter. Let me ask you for the file path or search for it in the context."
+- Retry with correct parameters
+
+Remember: Tool parameter validation errors should trigger immediate self-correction, not repeated failures.`;
+  }
+
+  private generateToolDescriptions(): string {
+    if (Object.keys(this.mcpClients).length === 0) {
+      return 'No MCP tools currently available.';
+    }
+
+    const descriptions: string[] = [];
+
+    for (const [serverName, client] of Object.entries(this.mcpClients)) {
+      const serverTools = client.getTools();
+      if (serverTools.length > 0) {
+        descriptions.push(`## ${serverName} server tools:`);
+
+        for (const tool of serverTools) {
+          let toolDescription = `- **${tool.name}**: ${
+            tool.description || 'No description available'
+          }`;
+
+          if (tool.inputSchema.required && tool.inputSchema.required.length > 0) {
+            toolDescription += `\n  - Required parameters: ${tool.inputSchema.required.join(', ')}`;
```

Comment:
the langgraph framework doesn't auto do this and requires us to manually build tool prompt?

Reply:
can check this and get back on it